File: app/src/main/java/com/kisssum/smartcity/ui/guidepages/GuideMainFragment.java
package com.kisssum.smartcity.ui.guidepages;

import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import androidx.viewpager2.widget.ViewPager2;

import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;

import com.kisssum.smartcity.R;
import com.kisssum.smartcity.databinding.FragmentGuideMainBinding;
import com.kisssum.smartcity.tool.API;
import com.kisssum.smartcity.tool.DecodeJson;

import java.io.IOException;
import java.lang.reflect.Array;
import java.util.ArrayList;

import okhttp3.FormBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

/**
 * A simple {@link Fragment} subclass.
 * Use the {@link GuideMainFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class GuideMainFragment extends Fragment {

    // TODO: Rename parameter arguments, choose names that match
    // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
    private static final String ARG_PARAM1 = "param1";
    private static final String ARG_PARAM2 = "param2";

    private FragmentGuideMainBinding binding;
    private Handler handler;

    // TODO: Rename and change types of parameters
    private String mParam1;
    private String mParam2;

    private ArrayList<String> imgUrls;

    public GuideMainFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param param1 Parameter 1.
     * @param param2 Parameter 2.
     * @return A new instance of fragment GuideMainFragment.
     */
    // TODO: Rename and change types and number of parameters
    public static GuideMainFragment newInstance(String param1, String param2) {
        GuideMainFragment fragment = new GuideMainFragment();
        Bundle args = new Bundle();
        args.putString(ARG_PARAM1, param1);
        args.putString(ARG_PARAM2, param2);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mParam1 = getArguments().getString(ARG_PARAM1);
            mParam2 = getArguments().getString(ARG_PARAM2);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentGuideMainBinding.inflate(inflater);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        SharedPreferences sp = requireActivity().getSharedPreferences("Settings", Context.MODE_PRIVATE);
        boolean isFirstEnter = sp.getBoolean("isFirstEnter", true);

        // 判断是否是第一次进入
        if (isFirstEnter) {
            handler = new Handler() {
                @Override
                public void handleMessage(@NonNull Message msg) {
                    super.handleMessage(msg);

                    imgUrls = (ArrayList<String>) msg.obj;
                    initViewPager();
                }
            };

            loadRotationImgs();
        } else {
            NavController controller = Navigation.findNavController(requireActivity(), R.id.fragment_main);
            controller.popBackStack();
            controller.navigate(R.id.navControlFragment);
        }
    }

    private void loadRotationImgs() {
        new Thread(() -> {
            OkHttpClient client = new OkHttpClient();

            Request request = new Request.Builder()
                    .url(API.INSTANCE.getGuideRotationListUrl(requireContext()))
                    .build();

            try {
                Response response = client.newCall(request).execute();
                String string = response.body().string();
                ArrayList<String> imgs = DecodeJson.INSTANCE.decodeGuideRotationList(string);

                Message message = new Message();
                message.obj = imgs;
                handler.sendMessage(message);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }).start();
    }

    private void initViewPager() {
        FragmentStateAdapter adapter = new FragmentStateAdapter(requireActivity()) {
            @NonNull
            @Override
            public Fragment createFragment(int position) {
                return new GuidePagesFragment(position, imgUrls.get(position));
            }

            @Override
            public int getItemCount() {
                return 5;
            }
        };

        binding.viewPager.setAdapter(adapter);
        binding.viewPager.getChildAt(0).setOverScrollMode(View.OVER_SCROLL_NEVER);
        binding.viewPager.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
            @Override
            public void onPageSelected(int position) {
                binding.guideCard0.setCardBackgroundColor(Color.GRAY);
                binding.guideCard1.setCardBackgroundColor(Color.GRAY);
                binding.guideCard2.setCardBackgroundColor(Color.GRAY);
                binding.guideCard3.setCardBackgroundColor(Color.GRAY);
                binding.guideCard4.setCardBackgroundColor(Color.GRAY);

                if (position == 0) binding.guideCard0.setCardBackgroundColor(Color.WHITE);
                else if (position == 1) binding.guideCard1.setCardBackgroundColor(Color.WHITE);
                else if (position == 2) binding.guideCard2.setCardBackgroundColor(Color.WHITE);
                else if (position == 3) binding.guideCard3.setCardBackgroundColor(Color.WHITE);
                else if (position == 4) binding.guideCard4.setCardBackgroundColor(Color.WHITE);
            }
        });

        binding.guideCard0.setOnClickListener(view1 -> binding.viewPager.setCurrentItem(0));
        binding.guideCard1.setOnClickListener(view1 -> binding.viewPager.setCurrentItem(1));
        binding.guideCard2.setOnClickListener(view1 -> binding.viewPager.setCurrentItem(2));
        binding.guideCard3.setOnClickListener(view1 -> binding.viewPager.setCurrentItem(3));
        binding.guideCard4.setOnClickListener(view1 -> binding.viewPager.setCurrentItem(4));
    }

    @Override
    public void onResume() {
        super.onResume();

        requireActivity().getWindow().addFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);
        requireActivity().getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();

        requireActivity().getWindow().clearFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);
    }
}

File: app/src/main/java/com/kisssum/smartcity/ui/guidepages/GuideMainFragment.kt
package com.kisssum.smartcity.ui.guidepages

import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.EditText
import android.widget.ImageView
import androidx.fragment.app.Fragment
import androidx.navigation.Navigation
import androidx.viewpager.widget.ViewPager
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley
import com.bumptech.glide.Glide
import com.kisssum.smartcity.R
import com.kisssum.smartcity.databinding.FragmentGuideMainBinding
import com.kisssum.smartcity.tool.API
import com.kisssum.smartcity.tool.DecodeJson
import com.youth.banner.loader.ImageLoader

/**
 * A simple [Fragment] subclass.
 * Use the [GuideMainFragment.newInstance] factory method to
 * create an instance of this fragment.
 */
class GuideMainFragment : Fragment() {
    private lateinit var binding: FragmentGuideMainBinding

    // TODO: Rename and change types of parameters
    private var mParam1: String? = null
    private var mParam2: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            mParam1 = arguments?.getString(ARG_PARAM1)
            mParam2 = arguments?.getString(ARG_PARAM2)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        binding = FragmentGuideMainBinding.inflate(inflater)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val sp = requireActivity().getSharedPreferences("Settings", Context.MODE_PRIVATE)
        val isFirstEnter = sp.getBoolean("isFirstEnter", true)

        // 判断是否是第一次进入
        if (isFirstEnter) {
            loadRotationImgs()
        } else {
            val controller = Navigation.findNavController(requireActivity(), R.id.fragment_main)
            controller.popBackStack()
            controller.navigate(R.id.navControlFragment)
        }
    }

    private fun loadRotationImgs() {
        val sp = requireActivity().getSharedPreferences("Settings", Context.MODE_PRIVATE)

        Volley.newRequestQueue(requireContext()).apply {
            val stirngRequest = StringRequest(
                    API.getGuideRotationListUrl(requireContext()),
                    {
                        val urls = DecodeJson.decodeGuideRotationList(it)

                        binding.guideBanner.apply {
                            this.setImageLoader(object : ImageLoader() {
                                override fun displayImage(context: Context?, path: Any?, imageView: ImageView?) {
                                    Glide.with(context!!).load(API.getBaseUrl(context) + path.toString()).into(imageView!!)
                                }
                            })
                            this.setImages(urls)
                            this.isAutoPlay(false)
                            this.start()
                            
                            this.setOnPageChangeListener(object : ViewPager.OnPageChangeListener {
                                override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {

                                }

                                override fun onPageSelected(position: Int) {
                                    if (position == urls.size - 1) {
                                        binding.guideNetwork.visibility = View.VISIBLE
                                        binding.guideGo.visibility = View.VISIBLE

                                        binding.guideGo.setOnClickListener { v ->
                                            // 保存
                                            sp.edit().putBoolean("isFirstEnter", false).apply()
                                            val controller = Navigation.findNavController(requireActivity(), R.id.fragment_main)
                                            controller.popBackStack()
                                            controller.navigate(R.id.navControlFragment)
                                        }

                                        binding.guideNetwork.setOnClickListener { v ->
                                            val view1 = layoutInflater.inflate(R.layout.alertdialog_change_ip, null)
                                            val ip = view1.findViewById<EditText>(R.id.ip)
                                            val duankou = view1.findViewById<EditText>(R.id.duankou)
                                            ip.setText(sp.getString("ip", ""))
                                            duankou.setText(sp.getString("duankou", ""))
                                            AlertDialog.Builder(requireContext())
                                                    .setTitle("网络设置")
                                                    .setView(view1)
                                                    .setPositiveButton("保存") { dialog: DialogInterface?, which: Int ->
                                                        sp.edit()
                                                                .putString("ip", ip.text.toString())
                                                                .putString("duankou", duankou.text.toString())
                                                                .apply()
                                                    }
                                                    .setNegativeButton("取消") { dialog: DialogInterface?, which: Int -> }
                                                    .create()
                                                    .show()
                                        }
                                    } else {
                                        binding.guideNetwork.visibility = View.INVISIBLE
                                        binding.guideGo.visibility = View.INVISIBLE
                                    }
                                }

                                override fun onPageScrollStateChanged(state: Int) {

                                }
                            })
                        }
                    },
                    {}
            )

            this.add(stirngRequest)
        }
    }

    override fun onResume() {
        super.onResume()
        requireActivity().window.addFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS)
        requireActivity().window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR
    }

    override fun onDestroyView() {
        super.onDestroyView()
        requireActivity().window.clearFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS)
    }

    companion object {
        // TODO: Rename parameter arguments, choose names that match
        // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
        private const val ARG_PARAM1 = "param1"
        private const val ARG_PARAM2 = "param2"

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @param param1 Parameter 1.
         * @param param2 Parameter 2.
         * @return A new instance of fragment GuideMainFragment.
         */
        // TODO: Rename and change types and number of parameters
        fun newInstance(param1: String?, param2: String?): GuideMainFragment {
            val fragment = GuideMainFragment()
            val args = Bundle()
            args.putString(ARG_PARAM1, param1)
            args.putString(ARG_PARAM2, param2)
            fragment.arguments = args
            return fragment
        }
    }
}